dao-contracts/tests/test_admin.rs

```diff
@@ -0,0 +1,125 @@
+
+            admin_contract.assert_last_event(
+                VotingContractCreated {
+                    variable_repo: Address::from(variable_repo_contract.get_package_hash()),
+                    reputation_token: Address::from(reputation_token_contract.get_package_hash()),
+                    voter_contract: Address::from(admin_contract.get_package_hash()),
+                },
+            );
+        }
+
+        context "voting" {
+            before {
+                admin_contract
+                .create_voting(
+                    reputation_token_contract.address(),
+                    Action::AddToWhitelist,
+                    env.get_account(1),
+                    minimum_reputation,
+                )
+                .unwrap();
+
+                let voting_id = VotingId::zero();
+                let voting: Voting = admin_contract.get_voting(voting_id);
+
+                // cast votes for informal voting
+                admin_contract
+                    .as_account(env.get_account(1))
+                    .vote(voting_id, true, U256::from(500))
+                    .unwrap();
+
+                // fast forward
+                env.advance_block_time_by(Duration::from_secs(voting.informal_voting_time() + 1));
+
+                // finish informal voting
+                admin_contract
+                    .as_account(env.get_account(1))
+                    .finish_voting(voting_id)
+                    .unwrap();
+
+                let voting_id = VotingId::one();
+
+                // cast votes for formal voting
+                admin_contract
+                    .as_account(env.get_account(1))
+                    .vote(voting_id, true, 1000.into())
+                    .unwrap();
+                admin_contract
+                    .as_account(env.get_account(2))
+                    .vote(voting_id, true, 1000.into())
+                    .unwrap();
+            }
+
+            test "action was not performed before finish" {
+                assert_eq!(reputation_token_contract.is_whitelisted(env.get_account(1)), false);
+            }
+
+            test "action was not performed on rejected voting" {
+                // vote against
+                admin_contract
+                    .as_account(env.get_account(3))
+                    .vote(voting_id, false, 5000.into())
+                    .unwrap();
+
+                env.advance_block_time_by(Duration::from_secs(voting.formal_voting_time() + 1));
+                admin_contract.finish_voting(voting_id).unwrap();
+                assert_eq!(reputation_token_contract.is_whitelisted(env.get_account(1)), false);
+            }
+
+            test "action was performed after finish" {
+                env.advance_block_time_by(Duration::from_secs(voting.formal_voting_time() + 1));
+                admin_contract.finish_voting(voting_id).unwrap();
+
+                assert!(reputation_token_contract.is_whitelisted(env.get_account(1)));
+            }
```

Comment:
What would you say to write a test like this:
using an additional context (or describe) to add more info
then setup test in the `before` section
add the actual test has just an assertion
So we've got 
`given` - the outer context (voting in this case)
`when` -  the inner context 
`then` - the assertion
```
context "voting has been finished" {
    before {
        env.advance_block_time_by(Duration::from_secs(voting.formal_voting_time() + 1));
        admin_contract.finish_voting(voting_id).unwrap();

    }
    test "that action was performed" {
        assert!(reputation_token_contract.is_whitelisted(env.get_account(1)));
    }
}
```

Reply:
This would add a context with one test - I created contexts when there was a common code for at least two tests.
Additionally, I limited myself somewhat to the contexts nesting, so the code could fit in one screen.